src/core/packet.lua

```diff
@@ -59,23 +60,28 @@ end
 -- Prepend data to the start of a packet.
 function prepend (p, ptr, len)
    assert(p.length + len <= max_payload, "packet payload overflow")
-   C.memmove(p.data + len, p.data, p.length) -- Move the existing payload
+   shiftright(p, len)
    ffi.copy(p.data, ptr, len)                -- Fill the gap
-   p.length = p.length + len
    return p
 end
 
 -- Move packet data to the left. This shortens the packet by dropping
 -- the header bytes at the front.
 function shiftleft (p, bytes)
-   C.memmove(p.data, p.data+bytes, p.length-bytes)
+   -- C.memmove(p.data, p.data+bytes, p.length-bytes)
+   p.data = p.data + bytes
    p.length = p.length - bytes
 end
 
 -- Move packet data to the right. This leaves length bytes of data
 -- at the beginning of the packet.
 function shiftright (p, bytes)
-   C.memmove(p.data + bytes, p.data, p.length)
+   if p.data - bytes >= p.data_ - C.PACKET_HEADROOM_SIZE then
```

Comment:
Perhaps there should be a range check here (p.length + bytes <= PACKET_PAYLOAD_SIZE ?)